File: sdks/python/a2x_client.py
"""
A2X Python Client SDK

A lightweight Python client for the A2X gateway REST API.
Supports program execution, entity listing, agent probing,
and webhook registration.

API Reference:
    POST /a2x/execute        — Execute a Σ∞ program
    GET  /a2x/entities       — List connected entities/agents
    GET  /a2x/entities/:id   — Get entity/agent details
    GET  /a2x/probe/:agent   — Probe agent state
    POST /a2x/webhook        — Register a webhook callback

Usage:
    from a2x_client import A2xClient

    client = A2xClient("http://localhost:8778", api_key="my-key")
    result = client.execute("⟦Σ∞⟧⟬I:✦ ∷ C:⟨sys⟩ ∷ D:⟘⟭")
    print(result.result)
"""

from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Any, Optional

import requests


# ── Data types ──────────────────────────────────────────────────────────────

@dataclass
class ExecuteResponse:
    """Result of executing a Σ∞ program via the gateway."""
    result: str
    execution_time_ms: int
    status: str  # "completed", "error", "timeout"


@dataclass
class EntityInfo:
    """Information about a connected entity or agent."""
    id: str
    entity_type: str
    display_name: str
    capabilities: list[str]


@dataclass
class ProbeResponse:
    """Snapshot of an agent's internal state."""
    agent_id: str
    state: str
    ip: Optional[int]
    world_graph_size: int
    memory_trace_length: int


@dataclass
class WebhookResponse:
    """Result of registering a webhook callback."""
    webhook_id: str


# ── Client ─────────────────────────────────────────────────────────────────

class A2xClient:
    """Client for the A2X gateway REST API.

    Args:
        base_url: Gateway base URL (e.g., "http://localhost:8778").
        api_key: Optional API key for authentication.
        timeout: Request timeout in seconds (default 30).
    """

    def __init__(
        self,
        base_url: str = "http://localhost:8778",
        api_key: Optional[str] = None,
        timeout: float = 30.0,
    ):
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.timeout = timeout
        self._session = requests.Session()
        if api_key:
            self._session.headers["X-A2X-Key"] = api_key

    # ── Program Execution ──────────────────────────────────────────────

    def execute(
        self,
        program: str,
        format: str = "sigma",
        timeout_ms: int = 5000,
    ) -> ExecuteResponse:
        """Execute a Σ∞ program via the gateway.

        Args:
            program: Σ∞ program source text.
            format: Program format ("sigma" or "omega").
            timeout_ms: Execution timeout in milliseconds.

        Returns:
            ExecuteResponse with result, timing, and status.

        Raises:
            requests.HTTPError: On HTTP errors (4xx, 5xx).
            requests.ConnectionError: If the gateway is unreachable.
        """
        url = f"{self.base_url}/a2x/execute"
        params = {}
        if self.api_key:
            params["api_key"] = self.api_key

        payload = {
            "program": program,
            "format": format,
            "timeout_ms": timeout_ms,
        }

        resp = self._session.post(
            url,
            json=payload,
            params=params,
            timeout=self.timeout,
        )
        resp.raise_for_status()
        data = resp.json()
        return ExecuteResponse(
            result=data["result"],
            execution_time_ms=data["execution_time_ms"],
            status=data["status"],
        )

    # ── Entity Discovery ───────────────────────────────────────────────

    def list_entities(self) -> list[EntityInfo]:
        """List all connected entities and agents.

        Returns:
            List of EntityInfo objects.

        Raises:
            requests.HTTPError: On HTTP errors.
        """
        url = f"{self.base_url}/a2x/entities"
        resp = self._session.get(url, timeout=self.timeout)
        resp.raise_for_status()
        data = resp.json()
        return [
            EntityInfo(
                id=e["id"],
                entity_type=e["entity_type"],
                display_name=e["display_name"],
                capabilities=e["capabilities"],
            )
            for e in data
        ]

    def get_entity(self, entity_id: str) -> EntityInfo:
        """Get details about a specific entity or agent.

        Args:
            entity_id: The entity ID to look up.

        Returns:
            EntityInfo for the requested entity.

        Raises:
            requests.HTTPError: On HTTP errors (404 if not found).
        """
        url = f"{self.base_url}/a2x/entities/{entity_id}"
        resp = self._session.get(url, timeout=self.timeout)
        resp.raise_for_status()
        data = resp.json()
        return EntityInfo(
            id=data["id"],
            entity_type=data["entity_type"],
            display_name=data["display_name"],
            capabilities=data["capabilities"],
        )

    # ── Agent Probing ──────────────────────────────────────────────────

    def probe_agent(self, agent_id: str) -> ProbeResponse:
        """Probe an agent's internal state.

        Args:
            agent_id: The agent ID to probe.

        Returns:
            ProbeResponse with state snapshot.

        Raises:
            requests.HTTPError: On HTTP errors (404 if agent not found, 403 if probing is denied).
        """
        url = f"{self.base_url}/a2x/probe/{agent_id}"
        resp = self._session.get(url, timeout=self.timeout)
        resp.raise_for_status()
        data = resp.json()
        return ProbeResponse(
            agent_id=data["agent_id"],
            state=data["state"],
            ip=data.get("ip"),
            world_graph_size=data["world_graph_size"],
            memory_trace_length=data["memory_trace_length"],
        )

    # ── Webhooks ───────────────────────────────────────────────────────

    def register_webhook(
        self,
        callback_url: str,
        filter_correlation_ids: Optional[list[int]] = None,
    ) -> WebhookResponse:
        """Register a webhook callback for program completion events.

        Args:
            callback_url: URL to POST results to.
            filter_correlation_ids: Optional list of correlation IDs to filter.

        Returns:
            WebhookResponse with the registered webhook ID.
        """
        url = f"{self.base_url}/a2x/webhook"
        payload: dict[str, Any] = {"url": callback_url}
        if filter_correlation_ids:
            payload["filter_correlation_ids"] = filter_correlation_ids

        resp = self._session.post(url, json=payload, timeout=self.timeout)
        resp.raise_for_status()
        data = resp.json()
        return WebhookResponse(webhook_id=data["webhook_id"])

    # ── Health Check ──────────────────────────────────────────────────

    def health(self) -> bool:
        """Check if the gateway is reachable.

        Returns:
            True if the gateway responds, False otherwise.
        """
        try:
            resp = self._session.get(
                f"{self.base_url}/a2x/entities",
                timeout=5.0,
            )
            return resp.ok
        except requests.RequestException:
            return False

    # ── Context Manager ───────────────────────────────────────────────

    def close(self):
        """Close the underlying HTTP session."""
        self._session.close()

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close()


# ── Convenience Functions ──────────────────────────────────────────────────

def execute(
    program: str,
    base_url: str = "http://localhost:8778",
    api_key: Optional[str] = None,
) -> ExecuteResponse:
    """Quick one-shot program execution.

    Args:
        program: Σ∞ program source text.
        base_url: Gateway base URL.
        api_key: Optional API key.

    Returns:
        ExecuteResponse.
    """
    with A2xClient(base_url, api_key) as client:
        return client.execute(program)

File: sdks/typescript/a2x-client.ts
// A2X TypeScript/JavaScript Client SDK
//
// A lightweight client for the A2X gateway REST API.
// Supports program execution, entity listing, agent probing,
// and webhook registration.
//
// Usage:
//   import { A2xClient } from "./a2x-client";
//
//   const client = new A2xClient("http://localhost:8778", "my-key");
//   const result = await client.execute("⟦Σ∞⟧⟬I:✦ ∷ C:⟨sys⟩ ∷ D:⟘⟭");
//   console.log(result.result);

// ── Data types ──────────────────────────────────────────────────────────────

/** Result of executing a Σ∞ program via the gateway. */
export interface ExecuteResponse {
  /** Result Σ∞ program text. */
  result: string;
  /** Execution time in milliseconds. */
  execution_time_ms: number;
  /** Status: "completed", "error", or "timeout". */
  status: string;
}

/** Information about a connected entity or agent. */
export interface EntityInfo {
  /** Unique entity identifier. */
  id: string;
  /** Entity type classification. */
  entity_type: string;
  /** Human-readable display name. */
  display_name: string;
  /** Capabilities this entity provides. */
  capabilities: string[];
}

/** Snapshot of an agent's internal state. */
export interface ProbeResponse {
  /** Agent identifier. */
  agent_id: string;
  /** Current agent state (e.g., "idle", "running"). */
  state: string;
  /** Instruction pointer (null if VM has no program loaded). */
  ip: number | null;
  /** Number of nodes in the WorldGraph. */
  world_graph_size: number;
  /** Number of entries in the MemoryTrace. */
  memory_trace_length: number;
}

/** Result of registering a webhook callback. */
export interface WebhookResponse {
  /** Registered webhook identifier. */
  webhook_id: string;
}

// ── Client ─────────────────────────────────────────────────────────────────

export class A2xClient {
  private baseUrl: string;
  private apiKey?: string;
  private timeout: number;

  /**
   * Create a new A2X gateway client.
   *
   * @param baseUrl - Gateway base URL (default: "http://localhost:8778")
   * @param apiKey - Optional API key for authentication
   * @param timeout - Request timeout in milliseconds (default: 30000)
   */
  constructor(
    baseUrl: string = "http://localhost:8778",
    apiKey?: string,
    timeout: number = 30000,
  ) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.apiKey = apiKey;
    this.timeout = timeout;
  }

  private async request<T>(
    method: string,
    path: string,
    body?: unknown,
    params?: Record<string, string>,
  ): Promise<T> {
    const url = new URL(path, this.baseUrl);
    if (params) {
      for (const [key, value] of Object.entries(params)) {
        url.searchParams.set(key, value);
      }
    }

    const headers: Record<string, string> = {
      "Content-Type": "application/json",
    };
    if (this.apiKey) {
      headers["X-A2X-Key"] = this.apiKey;
    }

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      const response = await fetch(url.toString(), {
        method,
        headers,
        body: body ? JSON.stringify(body) : undefined,
        signal: controller.signal,
      });

      if (!response.ok) {
        const errorBody = await response.text();
        throw new Error(
          `A2X API error ${response.status}: ${errorBody}`,
        );
      }

      return (await response.json()) as T;
    } finally {
      clearTimeout(timeoutId);
    }
  }

  // ── Program Execution ──────────────────────────────────────────────

  /**
   * Execute a Σ∞ program via the gateway.
   *
   * @param program - Σ∞ program source text
   * @param format - Program format ("sigma" or "omega", default "sigma")
   * @param timeoutMs - Execution timeout in milliseconds (default 5000)
   * @returns ExecuteResponse with result, timing, and status
   */
  async execute(
    program: string,
    format: string = "sigma",
    timeoutMs: number = 5000,
  ): Promise<ExecuteResponse> {
    const params: Record<string, string> = {};
    if (this.apiKey) {
      params["api_key"] = this.apiKey;
    }

    return this.request<ExecuteResponse>("POST", "/a2x/execute", {
      program,
      format,
      timeout_ms: timeoutMs,
    }, params);
  }

  // ── Entity Discovery ───────────────────────────────────────────────

  /**
   * List all connected entities and agents.
   *
   * @returns Array of EntityInfo objects
   */
  async listEntities(): Promise<EntityInfo[]> {
    return this.request<EntityInfo[]>("GET", "/a2x/entities");
  }

  /**
   * Get details about a specific entity or agent.
   *
   * @param entityId - The entity ID to look up
   * @returns EntityInfo for the requested entity
   */
  async getEntity(entityId: string): Promise<EntityInfo> {
    return this.request<EntityInfo>(
      "GET",
      `/a2x/entities/${encodeURIComponent(entityId)}`,
    );
  }

  // ── Agent Probing ──────────────────────────────────────────────────

  /**
   * Probe an agent's internal state.
   *
   * @param agentId - The agent ID to probe
   * @returns ProbeResponse with state snapshot
   */
  async probeAgent(agentId: string): Promise<ProbeResponse> {
    return this.request<ProbeResponse>(
      "GET",
      `/a2x/probe/${encodeURIComponent(agentId)}`,
    );
  }

  // ── Webhooks ───────────────────────────────────────────────────────

  /**
   * Register a webhook callback for program completion events.
   *
   * @param callbackUrl - URL to POST results to
   * @param correlationIds - Optional list of correlation IDs to filter
   * @returns WebhookResponse with the registered webhook ID
   */
  async registerWebhook(
    callbackUrl: string,
    correlationIds?: number[],
  ): Promise<WebhookResponse> {
    const body: Record<string, unknown> = { url: callbackUrl };
    if (correlationIds) {
      body["filter_correlation_ids"] = correlationIds;
    }

    return this.request<WebhookResponse>("POST", "/a2x/webhook", body);
  }

  // ── Health Check ──────────────────────────────────────────────────

  /**
   * Check if the gateway is reachable.
   *
   * @returns True if the gateway responds, false otherwise
   */
  async health(): Promise<boolean> {
    try {
      const resp = await fetch(`${this.baseUrl}/a2x/entities`, {
        method: "GET",
        signal: AbortSignal.timeout(5000),
      });
      return resp.ok;
    } catch {
      return false;
    }
  }
}

// ── Convenience Function ──────────────────────────────────────────────────

/**
 * Quick one-shot program execution.
 *
 * @param program - Σ∞ program source text
 * @param baseUrl - Gateway base URL (default: "http://localhost:8778")
 * @param apiKey - Optional API key
 * @returns ExecuteResponse
 */
export async function execute(
  program: string,
  baseUrl: string = "http://localhost:8778",
  apiKey?: string,
): Promise<ExecuteResponse> {
  const client = new A2xClient(baseUrl, apiKey);
  return client.execute(program);
}

export default A2xClient;
